nixos/modules/system/boot/loader/raspberrypi/firmware-builder.nix

```diff
@@ -0,0 +1,27 @@
+{ pkgs, lib, stdenv, substituteAll, coreutils, bash, raspberrypifw, version
+, ubootEnabled ? false }:
+
+let
+  uboot =
+    if version == 0 then
+      pkgs.ubootRaspberryPiZero
+    else if version == 1 then
+      pkgs.ubootRaspberryPi
+    else if version == 2 then
+      pkgs.ubootRaspberryPi2
+    else if version == 3 then
+      if stdenv.hostPlatform.isAarch64 then
+        pkgs.ubootRaspberryPi3_64bit
+      else
+        pkgs.ubootRaspberryPi3_32bit
+    else
+      throw "U-Boot is not yet supported on the Raspberry Pi 4.";
```

Comment:
This is the error message I saw. So probably this part has gone out of sync.

Reply:
Huh, but apparently this still shouldn't work. Now I'm really curious why this works on my current config (based on unstable)

Reply:
Like a BIOS update would, the update needs to be applied mindfully by the owner.

@Gerschtli wrote good instructions for doing the update:

 - https://nixos.wiki/wiki/NixOS_on_ARM/Raspberry_Pi_4#Updating_U-Boot
 - https://github.com/NixOS/nixpkgs/issues/82455#issuecomment-959797355

Hopefully we can figure out a good "user story" for updating the platform firmware, one that does not require making a plethora of platform-specific configurations in the options.

(Relatedly, [I wrote about planning for a better NixOS on ARM](https://discourse.nixos.org/t/planning-for-a-better-nixos-on-arm/15346), and one of the goals **is** to push managing the platform firmware outside of the direct OS update lifecycle, for all distros.)

Reply:
Good point, and thanks for the pointers!

I was actually wondering if I could use `nixos-install` to refresh the firmware (and `config.txt` along with it) ...

Reply:
Nope, `nixos-install` is only a fancy wrapper around `nixos-rebuild`. If it can't be done with `nixos-rebuild` it can't be done with `nixos-install`.

The reason the SD image has the *Platform Firmware* (U-Boot) available at the front of the disk is because it's being put there during the image build. The *Platform Firmware* partition for the Raspberry Pi is totally decoupled from the actual in-use system.